### PR TITLE
[scheduler cleanup phase 1]:Move scheduler cache interface and implementation to pkg/scheduler/in…

### LINKED ISSUE #68961: [scheduler cleanup phase 1]: Move scheduler cache interface and implementation to pkg/scheduler/internal/cache
**PLEASE COORDINATE WITH SIG-SCHEDULING BEFORE WORKING ON THIS ISSUE.** We want to avoid the situation where two people accidentally start work on the same fix.
Main tracking issue: https://github.com/kubernetes/kubernetes/issues/68951

The scheduler’s cache implementation is not meant to be exported as a golang library outside of the scheduler. It should be moved from `k8s.io/kubernetes/pkg/scheduler/cache` to `k8s.io/kubernetes/pkg/scheduler/internal/cache`.

There are some types in the current `cache` package, namely `NodeInfo`, which should remain where they are. These are actually used outside of the scheduler, and need to be decoupled from the cache implementation.

/sig scheduling
/kind cleanup
/help
/good-first-issue

### PR DESCRIPTION
**Pull request by wgliang**
…ternal/cache

**What this PR does / why we need it**:

The scheduler’s cache implementation is not meant to be exported as a golang library outside of the scheduler. It should be moved from k8s.io/kubernetes/pkg/scheduler/cache to k8s.io/kubernetes/pkg/scheduler/internal/cache.

There are some types in the current cache package, namely NodeInfo, which should remain where they are. And all named as node package.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #68961 

**Special notes for your reviewer**:

**Release note**:

```release-note
Move scheduler cache interface and implementation to pkg/scheduler/internal/cache.
```

### CODE REVIEW

**Comment by wgliang** [comment]
/assign @misterikkit @bsalamat
Ready for review, PTAL.

File: cmd/kubeadm/.import-restrictions
Comment (misterikkit) [inline cmd/kubeadm/.import-restrictions]: golang will not allow imports of `pkg/scheduler/internal` from outside `pkg/scheduler`. I think this line should be removed. (same for kubectl.)

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (misterikkit) [inline pkg/kubelet/cm/devicemanager/manager.go]: I think there's no need to change how other programs are using our `NodeInfo` type in this PR. It will be significantly easier to merge this if we do not make changes outside of scheduling code.

File: pkg/scheduler/cache/node_info.go
Comment (misterikkit) [inline pkg/scheduler/cache/node_info.go]: We should not modify the `NodeInfo` API in this PR. Also, by replacing functions like this with a direct access, we risk introducing nil panics into other components.

**Review by misterikkit — CHANGES_REQUESTED** [review]
A majority of the lines changed in this PR are due to `NodeInfo` API changes. I think that is beyond the scope of this cleanup issue. (e.g. `NodeInfo` should not come from a package called `cache`. But that will have to be changed in a different PR.)

File: cmd/kubeadm/.import-restrictions
Comment (wgliang) [reply cmd/kubeadm/.import-restrictions]: if remove it, some error will like:

errors in package "k8s.io/kubernetes/pkg/kubectl/cmd/testing":
import k8s.io/kubernetes/pkg/scheduler/internal/cache did not match any allowed prefix.

File: pkg/kubelet/cm/devicemanager/manager.go
Comment (wgliang) [reply pkg/kubelet/cm/devicemanager/manager.go]: If don't change NodeInfo, all the fields in the structure are private, and it is very cumbersome and complicated to modify and use externally.

Comment (wgliang) [reply pkg/kubelet/cm/devicemanager/manager.go]: For example, we need to implement a lot of Set*() functions.

File: cmd/kubeadm/.import-restrictions
Comment (misterikkit) [reply cmd/kubeadm/.import-restrictions]: OK - I'm fine leaving it as-is. The go compiler should enforce our intentions anyway.

**Comment by resouer** [comment]
Please don't forget to fix all the verify failures and make CI happy :-)

**Comment by wgliang** [comment]
@resouer Sure. :)

**Comment by wgliang** [comment]
Done! @misterikkit

File: pkg/scheduler/cache/node_info.go
Comment (wgliang) [inline pkg/scheduler/cache/node_info.go]: Generation is public and easy to change.
(In the future, other fields should also be public.The current revision is just to avoid changing the code of other parts as much as possible. Just as misterikkit said,https://github.com/kubernetes/kubernetes/pull/68968#discussion_r220752522)

File: pkg/scheduler/cache/node_info.go
Comment (wgliang) [inline pkg/scheduler/cache/node_info.go]: GetPodKey need be called by other packages.

File: pkg/scheduler/cache/node_info_test.go
Comment (wgliang) [inline pkg/scheduler/cache/node_info_test.go]: copy from `pkg/scheduler/internal/cache`.

File: pkg/scheduler/internal/cache/cache_test.go
Comment (wgliang) [inline pkg/scheduler/internal/cache/cache_test.go]: Ignore generation field should include expected.

File: pkg/scheduler/internal/cache/cache_test.go
Comment (wgliang) [inline pkg/scheduler/internal/cache/cache_test.go]: All updated by public function.

File: pkg/scheduler/cache/node_info.go
Comment (misterikkit) [reply pkg/scheduler/cache/node_info.go]: Yeah, We should add a task in phase 2 to come back and clean this up.

File: pkg/scheduler/cache/node_info_test.go
Comment (misterikkit) [reply pkg/scheduler/cache/node_info_test.go]: I've never noticed this interface before. Honestly, I would rather get rid of it and have `makeBasePod` take a `*testing.T` arg. Since it's copy-paste, I don't mind leaving it as-is for the current PR.

**Review by misterikkit — APPROVED** [review]
/lgtm

**Comment by wgliang** [comment]
/assign @deads2k @timothysc @bsalamat 
please take a look, for approve.This is a basic PR, it also has three follow up PRs.

File: pkg/scheduler/cache/node_tree.go
Comment (bsalamat) [inline pkg/scheduler/cache/node_tree.go]: Why don't we move node_tree.go to internal/cache?

Comment (wgliang) [reply pkg/scheduler/cache/node_tree.go]: @misterikkit said:
There are some types in the current cache package, namely NodeInfo, which should remain where they are. These are actually used outside of the scheduler, and need to be decoupled from the cache implementation.

Comment (bsalamat) [reply pkg/scheduler/cache/node_tree.go]: Yes, but NodeTree is a recent addition. I would be surprised if it's used outside of the scheduler.

Comment (wgliang) [reply pkg/scheduler/cache/node_tree.go]: You are right, NodeTree is suitable for moving to internal/cache.I'll do it.

**Comment by wgliang** [comment]
Moved NodeTree to `pkg/scheduler/internal/cache`
Please take a look again. @misterikkit @bsalamat

**Comment by misterikkit** [comment]
@wgliang the change since last LGTM is just a rebase, right?

Also, this needs approval for the changes to .import-restrictions. Do you need help finding approvers?

**Comment by wgliang** [comment]
yes,just a rebase.
If you can merge as soon as possible, then it would be better.

**Comment by wgliang** [comment]
#68960 #68959 #68958 All of them base on this PR. If the current PR is merged, the three PRs will be reviewed at the same time. So, I am waiting for this PR to be merged.

**Comment by wgliang** [comment]
/ping @misterikkit @deads2k @timothysc
for approve

**Comment by wgliang** [comment]
add 
/ping @bsalamat

File: pkg/scheduler/internal/cache/node_tree.go
Comment (bsalamat) [inline pkg/scheduler/internal/cache/node_tree.go]: Why are we exporting this? Looks like we are using it only in internal/cache.

Comment (wgliang) [reply pkg/scheduler/internal/cache/node_tree.go]: Done. Thanks.

File: pkg/scheduler/cache/node_info.go
Comment (misterikkit) [reply pkg/scheduler/cache/node_info.go]: @bsalamat pointed out that allowing others to write to this field could cause errors. Maybe it should be hidden behind a getter?

File: pkg/scheduler/cache/node_info.go
Comment (misterikkit) [inline pkg/scheduler/cache/node_info.go]: This needs a comment that it is only to be used for testing. I think we can get rid of this in a future cleanup.

**Review by misterikkit — APPROVED** [review]
/lgtm

**Comment by misterikkit** [comment]
over to @bsalamat or @k82cn to approve.

File: pkg/kubectl/.import-restrictions
Comment (soltysh) [inline pkg/kubectl/.import-restrictions]: Why is this needed?

Comment (misterikkit) [reply pkg/kubectl/.import-restrictions]: I think it's due to a transitive dependency. Obviously that package cannot be imported by kubectl, but this line was needed to make the tooling happy.

/cc @lavalamp

Comment (lavalamp) [reply pkg/kubectl/.import-restrictions]: Yeah, import boss doesn't understand about internal packages.

**Comment by Huang-Wei** [comment]
I think leaving the release note as NONE would be more appropriate as it's internal code refactoring.

File: pkg/kubectl/.import-restrictions
Comment (soltysh) [reply pkg/kubectl/.import-restrictions]: Sigh...

Comment (apelisse) [reply pkg/kubectl/.import-restrictions]: Since we can't do better ...
/approve

**Review by timothysc — APPROVED** [review]
/approve

**Comment by wgliang** [comment]
It’s awful and there is a conflict. I will rebase immediately.

**Comment by wgliang** [comment]
Just rebse. 
Need your last lgtm lable.
/ping @misterikkit

**Pull request merged**
